### PR TITLE
IMAP Multi-Mailbox Search (RFC 7377)

### PR DESCRIPTION
**Pull request by ksmurchison**
This needs a set of Cassandane tests, but I need to get familiar with IMAPTalk

### CODE REVIEW

**Review by elliefm — COMMENTED** [review]
Looking okay so far

**Comment by ksmurchison** [comment]
Tests added

**Review by elliefm — APPROVED** [review]
Tests look comprehensive, and I built and ran it and all passed

**Comment by ksmurchison** [comment]
@brong @rsto Do you concur with Ellie that this is good to go?

File: imap/index.c [hunk 1]
```diff
@@ -1913,7 +1913,22 @@ static void begin_esearch_response(struct index_state *state,
...
+
+        if (searchargs->filter) {
+            /* RFC 7377: 2.1
+             * Each ESEARCH response MUST contain the MAILBOX, TAG,
+             * and UIDVALIDITY correlators. */
+            char *extname = mboxname_to_external(index_mboxname(state),
+                                                 searchargs->namespace,
+                                                 searchargs->userid);
+            prot_printf(state->out, " MAILBOX \"%s\" UIDVALIDITY %d",
+
```
Comment (rsto) [inline imap/index.c:1926]: nit: empty line

File: imap/imapd.c [hunk 2]
```diff
@@ -6114,21 +6191,121 @@ static void cmd_search(char *tag, int usinguid)
...
+                        mbname_t *mbname = mbname_from_intname(intname);
+                        mrock.root_depth = strarray_size(mbname_boxes(mbname));
+                        mbname_free(&mbname);
+                    }
+
+                    mboxlist_mboxtree(intname, multisearch_cb, &mrock, flags);
+                }
+                break;
+            }
+            }
```
Comment (rsto) [inline imap/imapd.c:6287]: nit: indenting

File: imap/imapparse.c [hunk 3]
```diff
@@ -575,6 +575,131 @@ static int get_search_date(struct protstream *pin, time_t *start, time_t *end)
...
+    }
+
+    do {
+        c = getastring(pin, pout, &extname);
+        if (!buf_len(&extname)) goto bad;
+
+        intname = mboxname_from_external(buf_cstring(&extname),
+                                         searchargs->namespace,
+                                         searchargs->userid);
+        strarray_appendm(mboxes, intname);
```
Comment (rsto) [inline imap/imapparse.c:603]: this could crash later if intname is NULL (strarray_appendm does not check for NULL string)

File: imap/imapd.c [hunk 4]
```diff
@@ -6051,27 +6057,98 @@ static void cmd_store(char *tag, char *sequence, int usinguid)
...
+        /* Mailbox is already selected - use current index state */
+        state = imapd_index;
+    }
+    else if ((r = index_open(mbentry->name, &mrock->init, &state))) {
+        return r;
+    }
+
+    /* Use a fresh search_expr_t for each mailbox */
+    if (mrock->args->root) search_expr_free(mrock->args->root);
+    mrock->args->root = search_expr_duplicate(mrock->expr);
```
Comment (rsto) [inline imap/imapd.c:6110]: This leaks
```
[510503] =====> Instance::_check_valgrind_logs[1317] ==510534==    at 0x4849E4C: malloc (vg_replace_malloc.c:307)
[510503] =====> Instance::_check_valgrind_logs[1317] ==510534==    by 0x4BE017F: xmalloc (xmalloc.c:55)
[510503] =====> Instance::_check_valgrind_logs[1317] ==510534==    by 0x4BE01E3: xzmalloc (xmalloc.c:64)
[510503] =====> Instance::_check_valgrind_logs[1317] ==510534==    by 0x4961073: search_expr_new_nnodes (search_expr.c:124)
[510503] =====> Instance::_check_valgrind_logs[1317] ==510534==    by 0x496110F: search_expr_duplicate_nnodes (search_expr.c:241)
[510503] =====> Instance::_check_valgrind_logs[1317] ==510534==    by 0x496466B: search_expr_duplicate (search_expr.c:261)
[510503] =====> Instance::_check_valgrind_logs[1317] ==510534==    by 0x12979B: cmd_search (imapd.c:6200)
[510503] =====> Instance::_check_valgrind_logs[1317] ==510534==    by 0x139313: cmdloop (imapd.c:1536)
[510503] =====> Instance::_check_valgrind_logs[1317] ==510534==    by 0x13C1CF: service_main (imapd.c:1008)
[510503] =====> Instance::_check_valgrind_logs[1317] ==510534==    by 0x151C37: main (service.c:647)
```

**Review by rsto — CHANGES_REQUESTED** [review]
I just looked at RFC 7377 for the first time, but the tests and implementation seem to cover it well.
 
Apart from a memory leak I just found some nits.

File: imap/index.c (hunk 1 above)
Comment (ksmurchison) [reply imap/index.c]: Hmm.  Must've been a random keystroke.  Fixed

File: imap/imapparse.c (hunk 3 above)
Comment (ksmurchison) [reply imap/imapparse.c]: Fixed.

File: imap/imapd.c (hunk 4 above)
Comment (ksmurchison) [reply imap/imapd.c:6110]: Leak fixed

File: imap/imapd.c (hunk 2 above)
Comment (ksmurchison) [reply imap/imapd.c:6289]: @rsto This is an artifact of declaring variables inside a switch clause.  Its looks weird, but this is how Emacs does the indenting.  How would you like this to look>

**Review by brong — APPROVED** [review]
This looks good, thanks Ken.

I had some questions about cmd_cancelled but I think it's already called in the inner loop so that's OK.

**Review by rsto — APPROVED** [review]
LGTM

File: imap/imapd.c (hunk 2 above)
Comment (elliefm) [reply imap/imapd.c:6289]: I don't like the two `}`s at the same indentation level either, but I thought about this myself, and concluded that it's an artefact of our style placing the `case` labels at the same indent level as the `switch` itself.

If our `switch` style was like:

```
switch(foo) {
    case 'a':
        // something simple;
        break;
    case 'b': {
        // something complex
        break;
    }
}
```

-- that is, with the cases at their own indent level -- then this wouldn't be a problem.  But that would use up a lot of horizontal space on whitespace, leaving less for code, meaning more lines need to be wrapped, leading to more complicated wrapping solutions...

I suppose the variable(s) in question could be moved to the containing scope, removing the need for `{}` in the `case`, but as it is their lifetime is clear, even if the `}}` is jarring.  If they were up a level, one would then need to read that whole scope holding them in mind, even though they're only used briefly in that one place.

So while I don't really like it, I didn't have a clearly better suggestion, and still don't.

**Pull request merged**